catalog/lease: monitor rangefeed to ensure progress

fqazi opened the pull request:
Previously, the lease manager has no way of monitoring the range feed to
ensure that updates were being received. This could be problematic with
session based leasing, which relies on this mechanism to detect new
version. In the expiry based model this would have less severe
consequences and lead to slow schema change (since eventually the old
leases would expire). In the session based model the range feed needs to
be functional, otherwise schema changes will be blocked forever. To
address this, this patch adds logic to bring a node down if the
system.descriptor range feed becomes inactive or doesn't start.

Fixes: #125207

Release note (bug fix): Addressed a bug where schema changes could hang
if the lease range feed stopped getting updates.

fqazi committed:
catalog/lease: monitor rangefeed to ensure progress

Previously, the lease manager has no way of monitoring the range feed to
ensure that updates were being received. This could be problematic with
session based leasing, which relies on this mechanism to detect new
version. In the expiry based model this would have less severe
consequences and lead to slow schema change (since eventually the old
leases would expire). In the session based model the range feed needs to
be functional, otherwise schema changes will be blocked forever. To
address this, this patch adds logic to refresh / poll descriptors for
manually.

Fixes: #125207

Release note (bug fix): Addressed a bug where schema changes could hang
if the lease range feed stopped getting updates.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125243)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2OeuED41UIOrtNeiu:-O-2OeuED41UIOrtNeiv:b-iwxtok) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_state.go#L157)):*
> ```Go
> 	defer s.mu.Unlock()
> 	// The desc is replacing an existing one at the same version.
> 	// Note: For session based leasing we can violate this rule, if
> ```

super nit: i think the comma kind of changes the intended meaning of this sentence. would it be more clear as:
```
	// Note: For session based leasing we can violate this rule if
	// the session is expired since that invalidates the active lease.
```
___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2PQwg3I0Gs7gokyO4:-O-2PQwg3I0Gs7gokyO5:b-47ie91) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_state.go#L160)):*
> ```Go
> 	// the session is expired (it invalidates the active lease).
> 	if !s.mu.expiration.Less(expiration) && (s.mu.session == nil ||
> 		!s.mu.session.Expiration().Less(s.mu.expiration)) {
> ```

this change is unrelated to the rangefeed monitoring right? should we get it in a separate commit so we can backport it with less risk?
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2S1bC96Jmrrouy4wn:-O-2S1bC96Jmrrouy4wo:bns120v) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_version_state.go#L163)):*
> ```Go
> 		s.t.m.sessionBasedLeasingModeAtLeast(ctx, SessionBasedDrain) {
> 		sessionExpiry := s.mu.session.Expiration()
> 		// If the fixed expiration lasts longer use that, just to have a consistent
> ```

is this comment inverted? it looks like the code is actually saying "if the session expiration lasts longer use that"
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2Ry-h3hyrIKL8qxMB:-O-2Ry-h3hyrIKL8qxMC:b-47ie91) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_version_state.go#L166)):*
> ```Go
> 		// view of expirations. However, if the session has expired always pick up
> 		// that time.
> 		if expiration.Less(sessionExpiry) || sessionExpiry.Less(s.t.m.storage.clock.Now()) {
> ```

this change is unrelated to the rangefeed monitoring right? should we get it in a separate commit so we can backport it with less risk?
___
*[`pkg/sql/catalog/lease/lease.go` line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2Uh0ZEMe11-_YdjuF:-O-2Uh0ZEMe11-_YdjuG:b-tilmh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L80)):*
> ```Go
> 	settings.Fraction)
> 
> var LeaseMonitorRangeFeed = settings.RegisterBoolSetting(
> ```

nit: looks like this can be lower case and stay private 
___
*[`pkg/sql/catalog/lease/lease.go` line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2W6bM9S5rnDtH5b8q:-O-2W6bN1Mba8N6FDqHm:b-2rvzum) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1500)):*
> ```Go
> 		ctx, "lease", []roachpb.Span{descriptorTableSpan}, hlc.Timestamp{}, handleEvent,
> 		rangefeed.WithSystemTablePriority(),
> 		rangefeed.WithOnCheckpoint(handleCheckpoint),
> ```

do you know how frequently checkpoints are expected from a rangefeed? just curious to help determine what the default intervals should be here
___
*[`pkg/sql/catalog/lease/lease.go` line 1515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2WSKW9WmM4GtxcMkQ:-O-2WSKW9WmM4GtxcMkR:be4m72k) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1515)):*
> ```Go
> // getRangeFeedMonitorFrequency determines how long the range feed becomes silent
> // before we started treating it as an availability issue on the cluster.
> func (m *Manager) getRangeFeedMonitorFrequency() time.Duration {
> ```

it seems like this will return `max(40 seconds, 5 minutes)` by default. i wonder if that's a little quick to decide if there's an unavailability issue, but i'm not sure how to decide this. have you thought about that?
___
*[`pkg/sql/catalog/lease/lease.go` line 1569 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2WGIn1vDRBUp9v1ry:-O-2WGIn1vDRBUp9v1rz:bl2ngbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1569)):*
> ```Go
> 			case <-m.refreshAllLeases:
> 				m.refreshSomeLeases(ctx, true /*refreshAll*/)
> 			case <-rangeFeedProgressWatchDog.C:
> ```

as per the comment on this function, this function might be deleted entirely once we remove the old expiration based lease code. is there another background task where it might make sense to have this check?
___
*[`pkg/sql/catalog/lease/testutils.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2UABH2O2h4f08eUuV:-O-2UABH2O2h4f08eUuW:bmk29w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/testutils.go#L62)):*
> ```Go
> 
> 	// DisableRangeFeedCheckpoint
> 	DisableRangeFeedCheckpoint func() bool
> ```

can this just be a `bool` or `AtomicBool` rather than a function? all of the usages of it are setting it to a one line function that returns true or false.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
One day I'm going to figure out how to fully unsubscribe and check out... but until that day, here's my initial reaction without context: this feels dire.

I feel like there's two things to be sorted out:

1) Why does a rangefeed fail?
2) How does the system recover when it detects such failure?

This is addressing (2), but it's doing it in a drastic way.

On sql-only notes, I imagine this will result in a crash. On a system node, it will lead eventually to unavailability of all sql operations until that rangefeed comes back. Maybe that's sane, but likely it's not great. In particular, there's nothing telling the rangefeed that it should consider disconnecting from the follower it's connected to and trying the leaseholder or something like that, right?

My 2c is that a better than pausing liveness heartbeating, which is rather layer violating, would be to have the lease layer recover from the missing rangefeed data gracefully.

A simple approach would be to detect that the rangeFeedIsUnavailable as you do here, and then in the background re-acquire the leases. If you find there's a new version, yay, success! A version of this is to set the lease to be dropped when its refcount hits zero if the rangefeed is marked as unavailable and the lease was acquired some amount of time ago. These two things fit together. Now, I think you also need to do something to kick the rangefeed.

Today, the logic isn't really capable of gracefully dealing with a gap in rangefeed updates. I think that's the big gap that needs to be resolved: find a way to restart the rangefeed from "now" and gracefully recover back to a happy state.

One approach for that would be to background refresh leases which are old if the rangefeed is not up to date. If the descriptor has a new version and you need to drop the current one, then you'll discover it in the background refresh.

A downside of all of this, I suppose, is that it could lead to more load for refreshing leases, but it's not going to be worse than the node starting up afresh, right?

---

Concrete summary of a straw man to be picked apart:

* Keep the monitoring stuff you added
* Trigger the code to refresh when they are dereferenced if the rangefeed is down and the acquisition time is some definition of old.
     * Perhaps some bookkeeping is needed to know whether the descriptor has already been queued for refresh.
* To pair with the above logic, keep some background task to refresh or drop leases if the rangefeed is behind

Separately, maybe, actually restart the rangefeed at a new timestamp and keep track of that timestamp. If you have a descriptor which was acquired at a timestamp before the rangefeed started, then treat it like you treat the descriptors while the rangefeed is down -- you're unsure of whether they are up to date.

What do you think?

fqazi commented:
@ajwerner Yeah, that proposal makes sense we would need refreshes on both dereferences and background refreshes / lease clean up (for descriptors with a reference count of zero), when we enter this state. It will at least make the node available in the scenario while we attempt to get the range feed started again.

I'll probably add some type of logging for the degraded state to make our lives on the support side easier.

fqazi committed:
catalog/lease: add logic to restart the range feed

If the range feed is no longer checkpointing, we added a polling-based
mitigation strategy to ensure nothing bad happens. However, this could
get expensive since it will lead to more round trips in general. To
address this, this patch also adds logic that will attempt to restart
the change feed.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2OeuED41UIOrtNeiu:-O-9EnVv34CHXxrhLDb2:b-it18vy) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_state.go#L157)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: i think the comma kind of changes the intended meaning of this sentence. would it be more clear as:
```
	// Note: For session based leasing we can violate this rule if
	// the session is expired since that invalidates the active lease.
```
</blockquote></details>

Done.

I dropped this change .
___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2PQwg3I0Gs7gokyO4:-O-9EscWEX-JRxiuVpu2:b-5c93ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_state.go#L160)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this change is unrelated to the rangefeed monitoring right? should we get it in a separate commit so we can backport it with less risk?
</blockquote></details>

Done.

So, this was there originally because the session expiry is ignored outside of serverless in the deadline code. So, this was way of having it take effect at the leasing level, but thats no longer needed.
___
*[`pkg/sql/catalog/lease/lease.go` line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2Uh0ZEMe11-_YdjuF:-O-9FPhU459717fU5wfL:b7t8p32) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L80)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: looks like this can be lower case and stay private
</blockquote></details>

So, with the latest changes both settings are used in the lease_test package.
___
*[`pkg/sql/catalog/lease/lease.go` line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2W6bM9S5rnDtH5b8q:-O-9FUxV8025YLDK0Bqa:b8jpsdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1500)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you know how frequently checkpoints are expected from a rangefeed? just curious to help determine what the default intervals should be here
</blockquote></details>

I believe these coincide with the closed timestamp intervals, so every ~3 seconds
___
*[`pkg/sql/catalog/lease/lease.go` line 1515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2WSKW9WmM4GtxcMkQ:-O-9Feio6xEt8ZMAfbE8:bdcf4g8) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1515)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it seems like this will return `max(40 seconds, 5 minutes)` by default. i wonder if that's a little quick to decide if there's an unavailability issue, but i'm not sure how to decide this. have you thought about that?
</blockquote></details>

Done.

I simplified this down to a single setting, where 0 disables the monitoring. The timer still runs to refresh this setting, but does nothing else every minute.
___
*[`pkg/sql/catalog/lease/lease.go` line 1569 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2WGIn1vDRBUp9v1ry:-O-9Funm48Q7p75istro:bimbi0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1569)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

as per the comment on this function, this function might be deleted entirely once we remove the old expiration based lease code. is there another background task where it might make sense to have this check?
</blockquote></details>

Good point, maybe we should rename this to something more generic? We still need the periodic thing for evicting descriptor

Maybe call this: PeriodicallyRunLeasingTasks and add a descriptive comment with what this covers:
- Evicting expired leases
- Rangefeed progress detection
- Legacy lease renewals
___
*[`pkg/sql/catalog/lease/testutils.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2UABH2O2h4f08eUuV:-O-9GHwYCZSXUK31KfqL:b-f5j0rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/testutils.go#L62)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can this just be a `bool` or `AtomicBool` rather than a function? all of the usages of it are setting it to a one line function that returns true or false.
</blockquote></details>

I can't embed a AtomicBool or AtomicInt, so I switched this to a int64 which is atomically set. We just need some way of flipping the testing knob after start up.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2S1bC96Jmrrouy4wn:-O-9GV0PAabNgc34zgcg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_version_state.go#L163)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this comment inverted? it looks like the code is actually saying "if the session expiration lasts longer use that"
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2Ry-h3hyrIKL8qxMB:-O-9GVH9B9500u9XE-w9:b-t2xvjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/descriptor_version_state.go#L166)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this change is unrelated to the rangefeed monitoring right? should we get it in a separate commit so we can backport it with less risk?
</blockquote></details>

Done.

These have been reverted.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 2 of 7 files at r1, 3 of 5 files at r2, 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rafiss)

___
*[`pkg/sql/catalog/lease/lease.go` line 882 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-C38zR-P1Ckfy4beBY:-O-C38zR-P1Ckfy4beBZ:b31c6bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7323d0a11ad4fe37b6b5257f80e7f6b7dafaf1/pkg/sql/catalog/lease/lease.go#L882)):*
> ```Go
> 
> 		// rangeFeedIsUnavailableTime tracks the time when the unavailability
> 		// even was detected, Each the range feed timer detects no check points
> ```

nit: spelling in the comment



_Suggestion:_
```Go
// event was detected. Each time the range feed timer detects
```
___
*[`pkg/sql/catalog/lease/lease.go` line 886 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-C0TQKEd3aAjimVtXC:-O-C0TQKEd3aAjimVtXD:b6zzob8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7323d0a11ad4fe37b6b5257f80e7f6b7dafaf1/pkg/sql/catalog/lease/lease.go#L886)):*
> ```Go
> 		rangeFeedIsUnavailableTime time.Time
> 
> 		// rangeFeedIsUnavailable tracks if the range feed is currently gunavailable.
> ```

nit: spelling



_Suggestion:_
```Go
unavailable
```
___
*[`pkg/sql/catalog/lease/lease.go` line 1715 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CAyVr9qbQMoWiZvxD:-O-CAyVr9qbQMoWiZvxE:bhlk3jy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c9665694a71f2966866f9e5a21a5a89fc4f7ede/pkg/sql/catalog/lease/lease.go#L1715)):*
> ```Go
> 			if len(desc.mu.active.data) == 0 && atomic.LoadInt32(&desc.renewalInProgress) == 0 {
> 				func() {
> 					m.mu.Lock()
> ```

Are we violating any mutex order protocol? We already hold the mutex in `desc.mu`. Are there cases where we lock in the opposite order? `refreshSomeLeases` does it in the opposite order I think.
___
*[`pkg/sql/catalog/lease/lease.go` line 1742 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-C8Im62huShw8ZMWqV:-O-C8Im62huShw8ZMWqW:b-aq4ro1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c9665694a71f2966866f9e5a21a5a89fc4f7ede/pkg/sql/catalog/lease/lease.go#L1742)):*
> ```Go
> 					}
> 				}); err != nil {
> 				log.Infof(ctx, "didnt refresh descriptor: %d lease: %s", desc.id, err)
> ```

Any reason we can't flow this error up the stack? It seems like if we changed `maybeCleanupOldVersions` to return an error we could properly flow it back.
___
*[`pkg/sql/catalog/lease/lease.go` line 900 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CmKTgAyfkGE_vdgp2:-O-CmKTgAyfkGE_vdgp3:b-63kgim) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L900)):*
> ```Go
> 
> 		// rangeFeedIsUnavailable tracks if the range feed is currently gunavailable.
> 		rangeFeedIsUnavailable bool
> ```

Do we need a separate bool to track this? Couldn't we check if `rangeFeedIsUnavailableTime` is zero.
___
*[`pkg/sql/catalog/lease/lease.go` line 1510 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-D5YrzBw6i2hQkbpSu:-O-D5YrzBw6i2hQkbpSv:bwiipnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L1510)):*
> ```Go
> 			return
> 		}
> 		m.mu.rangeFeedCheckpoints += 1
> ```

Are there any cases where we don't do check points for legit reasons? For instance, maybe the node is idle and has had no table access. 
___
*[`pkg/sql/catalog/lease/lease.go` line 1669 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CllvG8JVWRcML5374:-O-CllvG8JVWRcML5375:bkdi2zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L1669)):*
> ```Go
> 		return staleLeases[i].id < staleLeases[j].id
> 	})
> 	return staleLeases, !m.mu.rangeFeedIsUnavailable
> ```

I think `m.mu.rangeFeedIsUnavailable` would be true here. It is usually set to true when `m.mu.rangeFeedIsUnavailableTime` is set. So, we return false for the second output parameter. All the other return statements in this function also return false for that parameter. Are there any that return true?
___
*[`pkg/sql/catalog/lease/lease_test.go` line 3665 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-D6iY52jnWDB7ffG36:-O-D6iY52jnWDB7ffG37:b-i7ep7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease_test.go#L3665)):*
> ```Go
> 	// On node 1 intentionally disable the rangefeed, so that the watch dog
> 	// detects a problem.
> 	srv.ApplicationLayer(1).LeaseManager().(*lease.Manager).TestingSetDisableRangeFeedCheckpointFn(true)
> ```

Do we need to do this disable before we read in the descriptor at 'SELECT * from t1'? Or does the order not matter.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/catalog/lease/lease.go` line 1715 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CAyVr9qbQMoWiZvxD:-O-Gt8HdBqkei9wcdr2U:b-55zwd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c9665694a71f2966866f9e5a21a5a89fc4f7ede/pkg/sql/catalog/lease/lease.go#L1715)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Are we violating any mutex order protocol? We already hold the mutex in `desc.mu`. Are there cases where we lock in the opposite order? `refreshSomeLeases` does it in the opposite order I think.
</blockquote></details>

Good catch, yeah this can lead to  a deadlock. I modified the logic to free in the correct order via the caller.
___
*[`pkg/sql/catalog/lease/lease.go` line 1742 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-C8Im62huShw8ZMWqV:-O-GvJx73b55cXMyMlN3:b-wjbkic) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c9665694a71f2966866f9e5a21a5a89fc4f7ede/pkg/sql/catalog/lease/lease.go#L1742)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Any reason we can't flow this error up the stack? It seems like if we changed `maybeCleanupOldVersions` to return an error we could properly flow it back.
</blockquote></details>

Yeah, we could flow it up the stack but in terms of handling we would at best still end up logging it. The general problem is that we can't do much here with the error, outside of just having it retry naturally after assuming its a transient error.
___
*[`pkg/sql/catalog/lease/lease.go` line 900 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CmKTgAyfkGE_vdgp2:-O-GpQ_wElXkeRwZczvQ:b-1pqhpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L900)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we need a separate bool to track this? Couldn't we check if `rangeFeedIsUnavailableTime` is zero.
</blockquote></details>

So, the boolean allows us to do one final pass before the range feed recovers. I'll make that clear here in the comment, but the scenario is:

1) The range feed goes down at some time t.
2) We refresh descriptors at t+1
3) The range feed recovers at time t + 5 and starts providing updates from that point (because we restarted it)
4) At time t +6 we run one last update to make sure we are all caught up

Updated the comment:

// rangeFeedIsUnavailable tracks if the range feed is currently unavailable.  
// Instead of just keying off the unavailable time, this flag allows us to  
// provide one last update after the range feed recovers, which will ensure  
// that no stale data exists.
___
*[`pkg/sql/catalog/lease/lease.go` line 1510 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-D5YrzBw6i2hQkbpSu:-O-GrkMc2QJsRhN2cRQM:b-n0lym9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L1510)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Are there any cases where we don't do check points for legit reasons? For instance, maybe the node is idle and has had no table access.
</blockquote></details>

So, this is linked to the closed timestamp mechanism on the ranges, so I don't think there is a case where we will stop getting these:

https://www.cockroachlabs.com/docs/stable/architecture/transaction-layer#closed-timestamps
___
*[`pkg/sql/catalog/lease/lease.go` line 1669 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CllvG8JVWRcML5374:-O-GseOAEi3-8_wQgP2E:b-lbeyf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L1669)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I think `m.mu.rangeFeedIsUnavailable` would be true here. It is usually set to true when `m.mu.rangeFeedIsUnavailableTime` is set. So, we return false for the second output parameter. All the other return statements in this function also return false for that parameter. Are there any that return true?
</blockquote></details>

So, we could have the range feed become available again, which will leave the time still set, so that we can refresh things one last time before disabling this mechanism. I added the comment:

// If the rangefeed becomes available again, we want to refresh  
// the stale leases one last time, so that we are all caught up.
___
*[`pkg/sql/catalog/lease/lease_test.go` line 3665 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-D6iY52jnWDB7ffG36:-O-Gxhcw4GiaQCsSkPGR:b-eywvpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease_test.go#L3665)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we need to do this disable before we read in the descriptor at 'SELECT * from t1'? Or does the order not matter.
</blockquote></details>

There aren't any downsides to disabling it earlier, since the SELECT will be the first time the lease will be tracked on that node.


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-:-O-HVUTN-3kJzicmMP1h:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/lease/lease.go` line 900 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-CmKTgAyfkGE_vdgp2:-O-HSt72-uG2CyWS-Ib9:bpfuaqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4cbc0e101f857e17418cb093a5df9dcf8515db2/pkg/sql/catalog/lease/lease.go#L900)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

So, the boolean allows us to do one final pass before the range feed recovers. I'll make that clear here in the comment, but the scenario is:

1) The range feed goes down at some time t.
2) We refresh descriptors at t+1
3) The range feed recovers at time t + 5 and starts providing updates from that point (because we restarted it)
4) At time t +6 we run one last update to make sure we are all caught up

Updated the comment:

// rangeFeedIsUnavailable tracks if the range feed is currently unavailable.  
// Instead of just keying off the unavailable time, this flag allows us to  
// provide one last update after the range feed recovers, which will ensure  
// that no stale data exists.
</blockquote></details>

Thanks that helped to clarify this.


<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss RFAL

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi and @spilchen)

___
*[`-- commits` line 12 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-vsVrL9PwXobhFKIXo):*
nit: the PR description no longer matches the commit message
___
*[`pkg/sql/catalog/lease/descriptor_state.go` line 40 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O0AvS44-DLARiZf1PiI:-O0AvS44-DLARiZf1PiJ:b4mbiny) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/descriptor_state.go#L40)):*
> ```Go
> 
> 	// lastRefresh of this state
> 	lastRefreshTime atomic.Int64
> ```

nit: should it just be a `time.Time`, and have this be protected by the mutex? also, the comment could use a few more words.
___
*[`pkg/sql/catalog/lease/lease.go` line 1569 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2WGIn1vDRBUp9v1ry:-O-vsHlJ2vMtr2GFxfjy:b-yd5bxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/lease.go#L1569)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Good point, maybe we should rename this to something more generic? We still need the periodic thing for evicting descriptor

Maybe call this: PeriodicallyRunLeasingTasks and add a descriptive comment with what this covers:
- Evicting expired leases
- Rangefeed progress detection
- Legacy lease renewals
</blockquote></details>

yes, let's do this rename and improve the comments. can you do that in a separate commit in this PR?
___
*[`pkg/sql/catalog/lease/lease.go` line 87 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-w9wWy4Mgrzhz7bn0Y:-O-w9wWy4Mgrzhz7bn0Z:b-s87xhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/lease.go#L87)):*
> ```Go
> 	"sql.catalog.descriptor_lease_monitor_range_feed.timeout",
> 	"if the leasing subsystem does not receive any checkpoints for this timeout,"+
> 		" then the node is brought down (a setting of 0 means off)",
> ```

is the behavior still to bring the node down?
___
*[`pkg/sql/catalog/lease/lease.go` line 725 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O0AvHuaALj_Xfedwsva:-O0AvHub7QnOVgULBKPd:b8bqqdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/lease.go#L725)):*
> ```Go
> 		// it was released.
> 		return m.storage.
> 			release(ctx, m.stopper, lease)
> ```

nit: whitespace
___
*[`pkg/sql/catalog/lease/lease.go` line 1374 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O0AwGS55aSNUXgl6-BO:-O0AwGS55aSNUXgl6-BP:bq5xeu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/lease.go#L1374)):*
> ```Go
> // RangefeedLeases is not active.
> func (m *Manager) RefreshLeases(ctx context.Context, s *stop.Stopper, db *kv.DB) {
> 	m.mu.Lock()
> ```

what's the purpose of taking the lock here? most of the code here is in an async task, so this lock won't really be held during that period.
___
*[`pkg/sql/catalog/lease/lease.go` line 1687 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O0AyTOTBcw2IxnpjwkP:-O0AyTOTBcw2IxnpjwkQ:b8jb4lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/lease.go#L1687)):*
> ```Go
> // our alternative mechanism will manually poll and refresh stale
> // leases.
> func (m *Manager) pollAndMaybeRefreshOldLeases(
> ```

more of a general comment here: i feel like this is adding quite a bit of extra complexity to the leasing system, and it's logic that by it's very nature is pretty hard to test since it only happens when other lower level things are more seriously broken. 

could you remind me again if it's possible to take the approach of trying to restart the rangefeed if we detect that it is not up?

perhaps if the rangefeed isn't working and can't be restarted, it would be even better to force a manual investigation rather than trying to paper over the issue.


___
*[`pkg/sql/catalog/lease/lease.go` line 1789 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O0Ay3KMApoAOEcQZp2O:-O0Ay3KMApoAOEcQZp2P:bk4uo8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/33e036aa699a3f96e0f99c5e596c1e16a2940c58/pkg/sql/catalog/lease/lease.go#L1789)):*
> ```Go
> // with the range feed and manually refreshes potentially staled descriptors
> // if that is the case.
> func (m *Manager) handleRangeFeedAvailability(ctx context.Context) error {
> ```

nit: should it be called `handleRangeFeedUnavailability`?
___
*[`pkg/sql/catalog/lease/testutils.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125243#-O-2UABH2O2h4f08eUuV:-O0AwpRQ-TKYrR_Td86v:b8i6tjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/623f0cb0241b7fbbaa92e1a555deba4824b6c543/pkg/sql/catalog/lease/testutils.go#L62)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I can't embed a AtomicBool or AtomicInt, so I switched this to a int64 which is atomically set. We just need some way of flipping the testing knob after start up.
</blockquote></details>

not sure i follow -- what do you mean by not being able to embed an AtomicBool?


<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss Closing this will re-open a new PR
